0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+internal runtime implementation of `set` values.
+
+By comparison, the equivalent `foreach` loop is:
+
+```dafny
+foreach i | i in s {
+  print i, "\n";
+}
+```
+
+The runtime implementation of this loop can use the native iteration features of the target language.
+
+Going further, it is better to avoid writing any imperative loop at all where possible. Sticking to immutable data and
+expressions rather than statements allows logic to be used in specifications as well as implementation,
+and reduces the burden on the Dafny verifier. The proposed sequence comprehension expression allows more
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in a set is simple using a sequence comprehension expression: `seq i | i in s`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## foreach Loops
+
+A `foreach` loop in Dafny resembles `for` or `foreach` loops in many other mainstream languages,
+but is substantially more general and expressive. A very simple example only looks slightly different than expected:
+
+```dafny
+foreach x | x in [1, 2, 3, 4, 5] {
+  print x, "\n";
+}
+```
+
+The feature supports much more sophisticated uses as well, however,
+including binding multiple variables at once and filtering:
+
+```dafny
+var myDoubleDeckerSeq: seq<seq<int>> := ...;
+foreach x, y | x in myDoubleDeckerSeq && y in x && y != 0 {
+  Process(y);
+}
+```
+
+The high-level syntax for a `foreach` loop reuses several existing concepts:
+
+```
+ForeachLoop ::=
+  "foreach"
+  QuantifierDomain
+  LoopSpec
+  Body
+```
+
+More specifically, this expands to:
+
+```
+ForeachLoop ::=
+  "foreach" 
+  Ident [":" Type] { "," Ident [":" Type] } [ "|" Expression ]
+  { InvariantDecl | ModifiesDecl | DecreasesDecl }
+  Body
+```
+
+A `foreach` loop closely resembles a `forall` statement, the key difference being that a `forall`
+loop executes its body once for every tuple of quantified values simultaneously, whereas a `foreach` loop
+executes its body once for each tuple of quantified values in sequence, one at a time.
+
+Similarly to set comprehensions or assign-such-that statements, the domain of a `foreach` loop
+must be enumerable. The following loop would produce the error `"the domain of a foreach loop must be enumerable,
+but Dafny's heuristics can't figure out how to produce an enumeration for 'x'"`.
+
+```dafny
+foreach x: real | 0.0 <= x < 1.0 {
+  ...
+}
+```
+
+The domain is allowed to be potentially infinite if the loop is used in a compiled context and an explicit `decreases` clause is provided.
+`decreases *` is permitted, in which the `foreach` loop may never terminate. Any other `decreases` clause can be provided
+to ensure the loop terminates even if the domain is potentially infinite. The following (slightly bizarre) example is legal:
+
+```dafny
+foreach x: nat 
+  decreases 10 - x
```

Comment:
Yup, this is what happens when you can't actually run the verifier to catch your mistakes. :) Thanks.

Reply:
What should be added is not an invariant, but a range expression. There would be two problems with an invariant `x <= 10`. One problem would be that this does not hold true of all values that the `foreach` statement will consider (since this example used a `foreach` loop with no range expression). The other problem is that the loop invariant of a `foreach` statement is not allowed to mention the index variable (this is also true of the `for` statement).

Reply:
I think I need a better example here - it was questionable in the first place and trying to patch it with an invariant made it worse.